### ISSUE TITLE
JWT external group is not created after a successfull login

### ISSUE THREAD

**Issue by ezreal92170** [hashicorp/vault-plugin-auth-jwt#17]
Hello,

I am using the docker image of vault: library/vault:1.0.2

My JWT contains at the root level : 
` "gitlab_group": [
    "148:Divison Sécurité et Maîtrise des risques",
    "81:escapegame",
    "6:innovation",
    "343:hackathon-LOS",
    "5:annuaire"
  ],
  "preferred_username": "tm8enk"`

An entity for me is created but i don't see any external groups
I created a role user with the following config
`  "data": {
    "bound_audiences": [
      "onyxia"
    ],
    "bound_cidrs": null,
    "bound_subject": "",
    "groups_claim": "gitlab_group",
    "groups_claim_delimiter_pattern": "",
    "max_ttl": 0,
    "num_uses": 0,
    "period": 0,
    "policies": [
      "default"
    ],
    "ttl": 3600,
    "user_claim": "preferred_username"
  }`

In the debug log i see that : 
`2019-01-16T22:05:47.325Z [DEBUG] identity: refreshing external group memberships: entity_id=91ead235-1800-deef-2787-d00ebca51efa group_aliases=["mount_accessor:"auth_jwt_c8d09a1c" name:"148:Divison S\303\251curit\303\251 et Ma\303\256trise des risques" ", "mount_accessor:"auth_jwt_c8d09a1c" name:"81:escapegame" ", "mount_accessor:"auth_jwt_c8d09a1c" name:"6:innovation" ", "mount_accessor:"auth_jwt_c8d09a1c" name:"343:hackathon-LOS" ", "mount_accessor:"auth_jwt_c8d09a1c" name:"5:annuaire" "]
2019-01-16T22:06:24.991Z [DEBUG] rollback: attempting rollback: path=auth/jwt/
2019-01-16T22:06:24.991Z [DEBUG] rollback: attempting rollback: path=identity/
2019-01-16T22:06:24.991Z [DEBUG] rollback: attempting rollback: path=cubbyhole/
2019-01-16T22:06:24.991Z [DEBUG] rollback: attempting rollback: path=secret/
2019-01-16T22:06:24.991Z [DEBUG] rollback: attempting rollback: path=sys/
2019-01-16T22:06:24.991Z [DEBUG] rollback: attempting rollback: path=auth/token/`
`

**Comment by sergeytrasko** [comment]
Seems we have the same problem.
I expect that when a login operation happens and there is a 'groups' claim then missing Vault groups and group aliases will be created automatically.
Is it correct assumption?

**Comment by ezreal92170** [comment]
I did the same guess...

**Comment by tstoermer** [comment]
Hi,

the mapping needs some configuration. You need to create an external Group with an alias matching the group name in the token that you want to be mapped. All users will then be assigned in Vault automatically, if they perform a login.
see [External vs Internal Groups](https://www.vaultproject.io/docs/secrets/identity/index.html#external-vs-internal-groups)
and the [Tutorial](https://learn.hashicorp.com/vault/identity-access-management/iam-identity)

Hope this helps. I just discovered it some days ago while testing configuration.

**Comment by sergeytrasko** [comment]
Here the question is slightly different - is there a way to create external identity groups **automatically** within the Vault?
It is clear that one can create identity groups beforehand and it will work. 
But imagine we have from hundreds to thousands of groups with unpredictable lifecycle. Manual management of them would be a maintenance hell.
For us it would be good if Vault could create such groups once it sees such group in JWT claim.

**Comment by jefferai** [comment]
@sergeytrasko The problem is the flip side: you want a defined subset of groups in Vault, but you authenticate and are a member of thousands of groups (this is a real thing, I promise you). Now you have thousands of groups that have to be managed within Vault, when you wanted three.

We may make this behavior toggleable at some point though.

**Comment by sergeytrasko** [comment]
@jefferai yes, I understand the behaviour you described. And I'm sure that 99% of Vault consumers would like to control which identity groups are available in Vault.

We have slightly different case - we want all groups to appear in Vault. These groups will have a certain naming pattern and we can use Sentinel policies to make decisions based on secret paths and group names. So for us it does not really make difference - three or thousands of groups.

For now we plan to go with some kind of ETL job that would create identity groups and aliases based on information from our OIDC server - probably this is the only option unless we have a toggle.

Another question would be - should it be a JWT-specific or a global toggle?

**Comment by jefferai** [comment]
Unsure, that's part of what we'd have to figure out. There's a third option, which is per-mount. Which may be the right granularity. Maybe? Possibly. Perhaps.

**Comment by ezreal92170** [comment]
I have the same usecase as @sergeytrasko . 
I would love if I can automacally bind some external group in the JWT claims and use the id or name of the group in policies in a secret engine.. I do it at the user level but i can't at a group level.
